src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointConstraintEndpointSelector.cs

```diff
@@ -32,37 +29,50 @@ internal class EndpointSelector
             _endpointConstraintCache = endpointConstraintCache;
         }
 
-        public Endpoint SelectBestCandidate(HttpContext context, IReadOnlyList<Endpoint> candidates)
+        public override Task SelectAsync(
+            HttpContext httpContext,
+            IEndpointFeature feature,
+            CandidateSet candidates)
```

Comment:
A quick work about this API definition. I passed in the `IEndpointFeature` for perf. Accessing the feature collection is somewhat expensive in the scale we're working in.

Also, this returns task. You're intended to modify the feature collection. The reason why this is important is that you need to set a few properties on the feature. An API where you just return an endpoint wouldn't be sufficient.